File: agentic-loops/agent_loop.ts
/**
 * Plug-and-play Agentic Loop — TypeScript
 * Requires: npm install @anthropic-ai/sdk
 * Set: ANTHROPIC_API_KEY env var
 */

import Anthropic from "@anthropic-ai/sdk";

const client = new Anthropic({ apiKey: process.env.ANTHROPIC_API_KEY });

// ── Define your tools here ────────────────────────────────────────────────────
const TOOLS: Anthropic.Tool[] = [
  {
    name: "get_weather",
    description: "Get the current weather for a city.",
    input_schema: {
      type: "object",
      properties: {
        city: { type: "string", description: "City name" },
      },
      required: ["city"],
    },
  },
];

function executeTool(name: string, input: Record<string, string>): string {
  if (name === "get_weather") {
    return `The weather in ${input.city} is 72°F and sunny.`; // stub
  }
  return "Unknown tool";
}

// ── The loop ──────────────────────────────────────────────────────────────────
async function runAgent(userMessage: string, maxIterations = 10): Promise<string> {
  const messages: Anthropic.MessageParam[] = [
    { role: "user", content: userMessage },
  ];

  for (let i = 0; i < maxIterations; i++) {
    const response = await client.messages.create({
      model: "claude-opus-4-6",
      max_tokens: 1024,
      tools: TOOLS,
      messages,
    });

    messages.push({ role: "assistant", content: response.content });

    if (response.stop_reason === "end_turn") {
      for (const block of response.content) {
        if (block.type === "text") return block.text;
      }
    }

    if (response.stop_reason === "tool_use") {
      const toolResults: Anthropic.ToolResultBlockParam[] = [];
      for (const block of response.content) {
        if (block.type === "tool_use") {
          const result = executeTool(block.name, block.input as Record<string, string>);
          console.log(`[Tool] ${block.name}(${JSON.stringify(block.input)}) → ${result}`);
          toolResults.push({ type: "tool_result", tool_use_id: block.id, content: result });
        }
      }
      messages.push({ role: "user", content: toolResults });
    }
  }

  return "Max iterations reached.";
}

(async () => {
  const result = await runAgent("What's the weather like in Tokyo?");
  console.log(result);
})();

File: agentic-loops/AgentLoop.java
// Plug-and-play Agentic Loop — Java
// Requires: Java 11+, add to pom.xml or gradle:
//   com.squareup.okhttp3:okhttp:4.12.0
//   com.fasterxml.jackson.core:jackson-databind:2.17.0
// Set: ANTHROPIC_API_KEY env var

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import okhttp3.*;

import java.util.ArrayList;
import java.util.List;

public class AgentLoop {

    static final String API_URL = "https://api.anthropic.com/v1/messages";
    static final String MODEL = "claude-opus-4-6";
    static final OkHttpClient HTTP = new OkHttpClient();
    static final ObjectMapper MAPPER = new ObjectMapper();

    // ── Execute tools here ────────────────────────────────────────────────────
    static String executeTool(String name, JsonNode input) {
        if ("get_weather".equals(name)) {
            String city = input.path("city").asText("unknown");
            return "The weather in " + city + " is 72°F and sunny."; // stub
        }
        return "Unknown tool";
    }

    // ── The loop ──────────────────────────────────────────────────────────────
    static String runAgent(String userMessage, int maxIterations) throws Exception {
        String apiKey = System.getenv("ANTHROPIC_API_KEY");

        // Build tools definition
        ArrayNode tools = MAPPER.createArrayNode();
        ObjectNode tool = MAPPER.createObjectNode();
        tool.put("name", "get_weather");
        tool.put("description", "Get the current weather for a city.");
        ObjectNode schema = MAPPER.createObjectNode();
        schema.put("type", "object");
        ObjectNode props = MAPPER.createObjectNode();
        ObjectNode cityProp = MAPPER.createObjectNode();
        cityProp.put("type", "string");
        cityProp.put("description", "City name");
        props.set("city", cityProp);
        schema.set("properties", props);
        ArrayNode required = MAPPER.createArrayNode();
        required.add("city");
        schema.set("required", required);
        tool.set("input_schema", schema);
        tools.add(tool);

        // Message history
        List<ObjectNode> messages = new ArrayList<>();
        ObjectNode userMsg = MAPPER.createObjectNode();
        userMsg.put("role", "user");
        userMsg.put("content", userMessage);
        messages.add(userMsg);

        for (int i = 0; i < maxIterations; i++) {
            // Build request body
            ObjectNode body = MAPPER.createObjectNode();
            body.put("model", MODEL);
            body.put("max_tokens", 1024);
            body.set("tools", tools);
            ArrayNode msgArray = MAPPER.createArrayNode();
            messages.forEach(msgArray::add);
            body.set("messages", msgArray);

            Request request = new Request.Builder()
                .url(API_URL)
                .header("x-api-key", apiKey)
                .header("anthropic-version", "2023-06-01")
                .header("content-type", "application/json")
                .post(RequestBody.create(MAPPER.writeValueAsBytes(body),
                    MediaType.get("application/json")))
                .build();

            JsonNode resp;
            try (Response response = HTTP.newCall(request).execute()) {
                resp = MAPPER.readTree(response.body().string());
            }

            String stopReason = resp.path("stop_reason").asText();
            JsonNode content = resp.path("content");

            // Append assistant message
            ObjectNode assistantMsg = MAPPER.createObjectNode();
            assistantMsg.put("role", "assistant");
            assistantMsg.set("content", content);
            messages.add(assistantMsg);

            if ("end_turn".equals(stopReason)) {
                for (JsonNode block : content) {
                    if ("text".equals(block.path("type").asText())) {
                        return block.path("text").asText();
                    }
                }
            }

            if ("tool_use".equals(stopReason)) {
                ArrayNode toolResults = MAPPER.createArrayNode();
                for (JsonNode block : content) {
                    if ("tool_use".equals(block.path("type").asText())) {
                        String name = block.path("name").asText();
                        JsonNode input = block.path("input");
                        String result = executeTool(name, input);
                        System.out.printf("[Tool] %s → %s%n", name, result);
                        ObjectNode tr = MAPPER.createObjectNode();
                        tr.put("type", "tool_result");
                        tr.put("tool_use_id", block.path("id").asText());
                        tr.put("content", result);
                        toolResults.add(tr);
                    }
                }
                ObjectNode toolMsg = MAPPER.createObjectNode();
                toolMsg.put("role", "user");
                toolMsg.set("content", toolResults);
                messages.add(toolMsg);
            }
        }

        return "Max iterations reached.";
    }

    public static void main(String[] args) throws Exception {
        System.out.println(runAgent("What's the weather like in Tokyo?", 10));
    }
}
